src/Nethermind/Nethermind.Optimism/CL/P2P/OptimismCLP2P.cs

```diff
@@ -67,38 +69,62 @@ public OptimismCLP2P(ulong chainId, string[] staticPeerList, ICLConfig config, A
             })
             .AddSingleton(new Settings())
             .BuildServiceProvider();
+
+        _mainLoopTask = new(async () =>
+        {
+            await MainLoop();
+        });
     }
 
-    private ulong _headPayloadNumber = 0;
-    private readonly SemaphoreSlim _semaphore = new(1);
+    private ulong _headPayloadNumber;
+    private readonly Channel<ExecutionPayloadV3> _blocksP2PMessageChannel = Channel.CreateBounded<ExecutionPayloadV3>(10); // for safety add capacity
+
     private async void OnMessage(byte[] msg)
     {
-        await _semaphore.WaitAsync();
         try
         {
             if (TryValidateAndDecodePayload(msg, out var payload))
             {
                 if (_logger.IsTrace) _logger.Trace($"Received payload prom p2p: {payload}");
+                await _blocksP2PMessageChannel.Writer.WriteAsync(payload, _cancellationTokenSource.Token);
+            }
+        }
+        catch (Exception e)
+        {
+            if (e is not OperationCanceledException && _logger.IsError) _logger.Error("Unhandled exception in Optimism CL P2P:", e);
+        }
+    }
+
+    private async Task MainLoop()
+    {
+        while (true)
```

Comment:
maybe `await foreach (ExecutionPayloadV3 payload in reader.ReadAllAsync())` instead?

Will we ever close the channel, shouldn't we close it when app close (we have a CancelationToken available for that)?